miden-tx/src/executor/mod.rs

```diff
@@ -0,0 +1,125 @@
+        &mut self.compiler
+    }
+
+    /// Fetches the data required to execute the transaction from the [DataStore], compiles the
+    /// transaction into a [Program] using the [TransactionComplier], and returns a
+    /// [PreparedTransaction].
+    ///
+    /// Errors:
+    /// - If required data can not be fetched from the [DataStore].
+    /// - If the transaction can not be compiled.
+    fn prepare_transaction(
+        &mut self,
+        account_id: AccountId,
+        block_ref: u32,
+        note_origins: Vec<(u32, NodeIndex)>,
+        tx_script: Option<ProgramAst>,
+    ) -> Result<PreparedTransaction, TransactionExecutorError> {
+        let account = self
+            .data_store
+            .get_account(account_id)
+            .map_err(TransactionExecutorError::FetchTransactionDataFailed)?;
+
+        let block_header = self
+            .data_store
+            .get_block_header(block_ref)
+            .map_err(TransactionExecutorError::FetchTransactionDataFailed)?;
+
+        let notes = note_origins
+            .into_iter()
+            .map(|(block_num, note_index)| {
+                self.data_store
+                    .get_note(block_num, note_index)
+                    .map_err(TransactionExecutorError::FetchTransactionDataFailed)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let block_chain = self
+            .data_store
+            .get_chain_mmr(block_ref)
+            .map_err(TransactionExecutorError::FetchTransactionDataFailed)?;
+
+        let (tx_script_root, tx_program) = self
+            .compiler
+            .compile_transaction(account_id, &notes, tx_script)
+            .map_err(TransactionExecutorError::CompileTransactionError)?;
+
+        Ok(PreparedTransaction::new(
+            account,
+            block_header,
+            block_chain,
+            notes,
+            tx_script_root,
+            tx_program,
+        ))
+    }
+
+    /// Constructs a transaction from the provided inputs, executes it, and returns a [TxWitness]
+    /// which can be used to generate a proof for the transaction.
+    pub fn execute_transaction(
+        &mut self,
+        account_id: AccountId,
+        block_ref: u32,
+        note_origins: Vec<(u32, NodeIndex)>,
+        tx_script: Option<ProgramAst>,
+    ) -> Result<TxWitness, TransactionExecutorError> {
+        let transaction =
+            self.prepare_transaction(account_id, block_ref, note_origins, tx_script)?;
+
+        let advice_recorder: AdviceRecorder = transaction.advice_provider_inputs().into();
+        // TODO: Process is currently behind the `internals` feature flag
+        let mut process = Process::new(
+            transaction.tx_program().kernel().clone(),
+            transaction.stack_inputs(),
+            advice_recorder,
+        );
+        let _result = process
+            .execute(transaction.tx_program())
+            .map_err(TransactionExecutorError::ExecuteTransactionProgramFailed)?;
```

Comment:
Are we not able to use `processor::execute()` function here?

Reply:
The `execute()` function consumes the advice provider.  We need to maintain ownership of the advice provider in this functions scope so we can convert it to an advice proof (`AdviceInputs`).

Reply:
I believe it is possible to pass in a mutable reference into `execute()` function instead of transferring ownership (see some more explanation [here](https://github.com/0xPolygonMiden/miden-vm/blob/next/processor/src/advice/mod.rs#L46)).

Reply:
Or, I think more relevant is that we provide [this](https://github.com/0xPolygonMiden/miden-vm/blob/next/processor/src/advice/mod.rs#L190) implementation.

Reply:
Nice! That should work. Thanks for the heads up, I'll give it a go.